ch390/src/esp_eth_mac_ch390.c

```diff
@@ -0,0 +1,897 @@
+        gpio_isr_handler_add(emac->int_gpio_num, ch390_isr_handler, emac);
+    }
+    ESP_GOTO_ON_ERROR(eth->on_state_changed(eth, ETH_STATE_LLINIT, NULL), err, TAG, "lowlevel init failed");
+    /* reset ch390 */
+    ESP_GOTO_ON_ERROR(ch390_reset(emac), err, TAG, "reset ch390 failed");
+    /* verify chip id */
+    ESP_GOTO_ON_ERROR(ch390_verify_id(emac), err, TAG, "vefiry chip ID failed");
+    /* default setup of internal registers */
+    ESP_GOTO_ON_ERROR(ch390_setup_default(emac), err, TAG, "ch390 default setup failed");
+    /* clear multicast hash table */
+    ESP_GOTO_ON_ERROR(ch390_clear_multicast_table(emac), err, TAG, "clear multicast table failed");
+    /* get emac address from eeprom */
+    ESP_GOTO_ON_ERROR(ch390_get_mac_addr(emac), err, TAG, "fetch ethernet mac address failed");
+    return ESP_OK;
+err:
+    if (emac->int_gpio_num >= 0) {
+        gpio_isr_handler_remove(emac->int_gpio_num);
+        gpio_reset_pin(emac->int_gpio_num);
+    }
+    eth->on_state_changed(eth, ETH_STATE_DEINIT, NULL);
+    return ret;
+}
+
+static esp_err_t emac_ch390_deinit(esp_eth_mac_t *mac)
+{
+    emac_ch390_t *emac = __containerof(mac, emac_ch390_t, parent);
+    esp_eth_mediator_t *eth = emac->eth;
+    mac->stop(mac);
+    if (emac->int_gpio_num >= 0) {
+        gpio_isr_handler_remove(emac->int_gpio_num);
+        gpio_reset_pin(emac->int_gpio_num);
+    }
+    if (emac->poll_timer && esp_timer_is_active(emac->poll_timer)) {
+        esp_timer_stop(emac->poll_timer);
+    }
+    eth->on_state_changed(eth, ETH_STATE_DEINIT, NULL);
+    return ESP_OK;
+}
+
+static void emac_ch390_task(void *arg)
+{
+    emac_ch390_t *emac = (emac_ch390_t *)arg;
+    uint8_t status = 0;
+    uint8_t *cache;
+    while (1) {
+        // check if the task receives any notification
+        if (emac->int_gpio_num >= 0) {                                   // if in interrupt mode
+            if (ulTaskNotifyTake(pdTRUE, pdMS_TO_TICKS(1000)) == 0 &&   // if no notification ...
+                    gpio_get_level(emac->int_gpio_num) == 0) {               // ...and no interrupt asserted
+                continue;                                                // -> just continue to check again
+            }
+        } else {
+            ulTaskNotifyTake(pdTRUE, portMAX_DELAY);
+        }
+
+        /* clear interrupt status */
+        ch390_register_read(emac, CH390_ISR, &status);
+        ch390_register_write(emac, CH390_ISR, status);
+        /* packet received */
+        if (status & ISR_PR) {
+            do {
+                if (emac->parent.receive(&emac->parent, emac->rx_buffer, &emac->rx_len) == ESP_OK) {
+                    if (emac->rx_len == 0) {
+                        break;
+                    } else {
+                        ESP_LOGD(TAG, "receive len=%lu", emac->rx_len);
+                        /* pass the buffer to stack (e.g. TCP/IP layer) */
+                        cache = malloc(emac->rx_len);
```

Comment:
You need to check if memory allocation was successful. If not, you need to perform appropriate error handling. That was the purpose of original code you removed. The buffer length was not read twice when allocation was performed with `DM9051_ETH_MAC_RX_BUF_SIZE_AUTO` parameter as you mentioned in you original PR. I admit the original DM9051 code may not be super readable... It's been done that way to mimic the internal EMAC implementation where it makes more sense.

Reply:
As U wish~
```
        /* allocate memory and check whether allocation failed */
        cache = malloc(emac->rx_len);
        if (cache == NULL) {
            ESP_LOGE(TAG, "no memory for receive buffer");
            continue;
        }
```